feat: mark items with a tag

hojerst opened the pull request:
Hi!

I implemented the idea you mentioned in my last contribution: marking every item with a tag (I decided to hardcode this to `git-credential-1password`, but we might add a option to override this as well. Unfortunately `op item get` doesn't allow to filter out by tag. I added a `op item list --tags $marker` and filter the result in go by title. As we have the url in the `op item list` as well, we might as well ignore the title and filter out by the `urls` field instead, which is probably better.

Let me know what you think about the url thing and I will refactor this. (note: this shouldn't be a breaking change, as the url is set in previously generated items as well, we just ignore the title for fetching).

As this whole functionality blew up the code, I decided to split the code into one file per concern. (still everything in main package though).

Furthermore I don't print any error if no item is found for the host (but errors are printed if there are technical problems otherwise)

ethrgeist commented:
Hey @hojerst,

thanks again for the contribution.

My initial label idea was simpler, i'd simply didn't want to touch items with `erase` / `store` that did not have the label.
An additional failsafe to prevent the helper from eating credentials.

But i guess we can use it for filtering as well, but i wonder if it would be better to use the title, since there might be scenarios where i have different credentials for the Weblogin and for the git.

The README should probably have a simple flow chart for people to understand, how entries are matchedf/updated, if we are moving away from the "simple" titel match and take labels and url into account. This could also help them prepare their 1Password items accordingly.

Splitting into multiple files is fine, i quite like the [advocated layout from here](https://github.com/golang-standards/project-layout) for go projects.

Since this has not the claim to be a stable library to be used externally, i think it would be fine to move all files to `internal/`

BR!

hojerst commented:
Regarding the title discussion: I see your point regarding the Weblogin and git. That was one of the reasons, I implemented the Title prefix (the other one was, that I personally just tend to prefix my items anyway). However searching for the Host by URL field instead of the title shoudn't be a problem in the case of multiple entries, as we would only search inside of the tagged entries. If the user has two items for github.com, one with a token, one with his actual password and otp, the user could either just enter his token once and the helper would create a matching item with the tag set, or the user could pregenerate the item and tag it with `git-cerdential-1password`. I personally would advocate for using the url field for finding the right item (in the tagged set of items) and leaving the title untouched after initial creation. So the user can rename the item without problems.

Regarding the restructuring into packages/folders: I initially tried that and realised, that this would also mean we have to refactor a bit more because of the global variables in main with program state. could be a good thing though. I'll probably give it a shot this evening (CET).

ethrgeist commented:
> the user could either just enter his token once and the helper would create a matching item with the tag set, or the user could pregenerate the item and tag it with git-cerdential-1password. I personally would advocate for using the url field for finding the right item (in the tagged set of items) and leaving the title untouched after initial creation.

Yes i agree, using the label and url matching is a good idea and leave the name based organizing to the user. :)

hojerst commented:
I implemented finding by url

TODO: folder structure refatcoring and describing in README how the process works

hojerst commented:
Hello again,

as suggested, I moved most of the code to the internal package and added a section to the README which describes how items are selected/updated. Please let me know if you prefer further adjustments, otherwise i consider the change done from my end.

ethrgeist commented:
Hi @hojerst,

real life happened, now i had the chance to look at it.

Thanks for your work, i'll merge this one. :)

All the best
